ukis_pysat/raster.py

```diff
@@ -40,30 +40,37 @@ def __init__(self, data, dimorder="first", crs=None, transform=None):
             self.dimorder = dimorder
         else:
             raise TypeError("dimorder for bands or channels must be either 'first' or 'last'.")
-        try:
+
+        if isinstance(data, rasterio.io.DatasetReader):
+            self.dataset = data
+            self.crs = self.dataset.crs
+            self.transform = self.dataset.transform
+            self.__arr = self.dataset.read()
+
+        elif isinstance(data, str):
             self.dataset = rasterio.open(data)
             self.crs = self.dataset.crs
             self.transform = self.dataset.transform
             self.__arr = self.dataset.read()
-        except (TypeError, AttributeError):
-            if isinstance(data, rasterio.io.DatasetReader):
-                self.dataset = data
-                self.crs = self.dataset.crs
-                self.transform = self.dataset.transform
-                self.__arr = self.dataset.read()
-            elif isinstance(data, np.ndarray):
-                if crs is None:
-                    raise TypeError("if dataset is of type np.ndarray crs must not be None")
-                if transform is None:
-                    raise TypeError("if dataset is of type np.ndarray transform must not be None")
-                meta = {"dtype": data.dtype, "count": data.shape[0], "crs": crs, "driver": "GTiff"}
+
+        elif isinstance(data, np.ndarray):
+            if crs is None:
+                raise TypeError("if dataset is of type np.ndarray crs must not be None")
+            if transform is None:
+                raise TypeError("if dataset is of type np.ndarray transform must not be None")
+            if dimorder == "first":
                 self.__arr = data
-                self.transform = transform
-                self.dataset = self.__update_dataset(meta).open()
-                self.crs = self.dataset.crs
-                self.transform = self.dataset.transform
+            elif dimorder == "last":
+                self.__arr = reshape_as_raster(data)
             else:
-                raise TypeError("dataset must be of type str, rasterio.io.DatasetReader or np.ndarray")
+                raise AttributeError("dimorder for bands or channels must be either 'first' or 'last'.")

```

Comment:
I think this should be a TypeError, also it should never reach that (you could use `else` in 63), the check is done earlier in 39ff.

Reply:
Correct, thanks